store/postgres/src/mapping.rs

```diff
@@ -0,0 +1,1081 @@
+use diesel::connection::SimpleConnection;
+use diesel::{debug_query, PgConnection, RunQueryDsl};
+use graphql_parser::query as q;
+use graphql_parser::schema as s;
+use inflector::Inflector;
+use std::cmp::PartialOrd;
+use std::collections::{HashMap, HashSet};
+use std::fmt::{self, Write};
+use std::rc::Rc;
+use std::str::FromStr;
+
+use crate::mapping_sql::{
+    ClampRangeQuery, ConflictingEntityQuery, CopyQuery, EntityData, FilterQuery, FindQuery,
+    InsertQuery, RevertClampQuery, RevertRemoveQuery,
+};
+use graph::prelude::{
+    format_err, Entity, EntityChange, EntityChangeOperation, EntityFilter, EntityKey,
+    QueryExecutionError, StoreError, StoreEvent, SubgraphDeploymentId, ValueType,
+};
+
+use crate::block_range::BlockNumber;
+use crate::entities::STRING_PREFIX_SIZE;
+
+/// A string we use as a SQL name for a table or column. The important thing
+/// is that SQL names are snake cased. Using this type makes it easier to
+/// spot cases where we use a GraphQL name like 'bigThing' when we should
+/// really use the SQL version 'big_thing'
+///
+/// We use `SqlName` for example for table and column names, and feed these
+/// directly to Postgres. Postgres truncates names to 63 characters; if
+/// users have GraphQL type names that do not differ in the first
+/// 63 characters after snakecasing, schema creation will fail because, to
+/// Postgres, we would create the same table twice. We consider this case
+/// to be pathological and so unlikely in practice that we do not try to work
+/// around it in the application.
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub struct SqlName(String);
+
+impl SqlName {
+    pub fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+impl From<&str> for SqlName {
+    fn from(name: &str) -> Self {
+        SqlName(name.to_snake_case())
+    }
+}
+
+impl From<String> for SqlName {
+    fn from(name: String) -> Self {
+        SqlName(name.to_snake_case())
+    }
+}
+
+impl fmt::Display for SqlName {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+/// The SQL type to use for GraphQL ID properties. We support
+/// strings and byte arrays
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum IdType {
+    String,
+    #[allow(dead_code)]
+    Bytes,
+}
+
+#[derive(Debug, Clone)]
+pub struct Mapping {
```

Comment:
We agreed in our meeting that this should be renamed to `Layout`; I will do that as the very last thing in this review since it will be a messy commit. Other things that should be renamed:

- `mapping` module -> `relational` module
- `mapping_sql` module -> `relational_queries` module